README
README for MIDI-Perl
                                        Time-stamp: "2005-01-29 15:27:16 AST"

			      MIDI-Perl

MIDI-Perl is a suite of Perl modules that allows you to read, compose,
modify, and write MIDI files.  It is largely object-oriented, but if
you follow the example code, you should be able to get along fine.

See http://interglacial.comu/~sburke/pub/perl_midi/ for more info.



PREREQUISITES

This suite requires Perl 5.

MIDI-Perl uses no external libraries at this time, aside from Carp,
which has been standard for as long as I can remember.  If you want to
use MIDI::Opus::draw, you'll also need GD.pm, however.



INSTALLATION

You install this module-suite, as you would install any perl module
library, by running these commands:

   perl Makefile.PL
   make
   make test
   make install

If you want to install a private copy of this module-suite in your home
directory, then you should try to produce the initial Makefile with
something like this command:

  perl Makefile.PL PREFIX=~/perl

See perldoc perlmodinstall for more information on installing modules.



DOCUMENTATION

See lib/MIDI.pm for an overview of the suite.  See ChangeLog for
recent changes.  POD-style documentation is included in all modules.
POD is readable with the 'perldoc' utility.



MACPERL INSTALLATION NOTES

Don't bother with the makefiles.
Just move MIDI.pm and the "MIDI" folder into your MacPerl library
folder.  Read the PODs with Shuck.



SUPPORT

Questions, bug reports, useful code bits, and suggestions for
MIDI-Perl should just be sent to me at sburke@cpan.org



AVAILABILITY

The latest version of MIDI-Perl is available from the Comprehensive
Perl Archive Network (CPAN).  <http://search.cpan.org/> is my favorite
way to view CPAN.



COPYRIGHT

Copyright 1998+ by Sean M. Burke <sburke@cpan.org.


ChangeLog

Revision history for Perl suite MIDI-Perl

        
                                Time-stamp: "2010-02-21 12:55:15 conklin"

2010-02-14  Darrell Conklin  conklin@cpan.org

	* maintainer -- DC has taken over as maintainer.  Thanks to SB for
	his excellent module.

	* Release 0.82
	
	* bug fix in Score.pm where events_r_to_score_r would destructively modify events

	* bug fix in Score.pm, when presented with two simultaneous events
	with the same note/channel, which led to unpredictable durations
	of score events, sometimes negative.
	
	* added grid quantization (to Opus, and Score)

2005-01-29  Sean M. Burke  sburke@cpan.org

	* Release 0.81 -- maintenance release.

	* No Perl changes -- just doc tweaks mostly.
	
	* Rearranged the dist's contents to be more modern-like.

	* Of course, more cowbell.
	
	
2002-08-21  Sean M. Burke  sburke@cpan.org

	* Release 0.80 -- just a docfix release, basically

	* Removing the out-dated Filespec.pod file.

	* To work around vstrings in Perl 5.6 (etc), MIDI::Simple volume
	specifications can now be expressed as V64 (etc), whereas
	previously you could only do v64.
	Bumping MIDI::SImple version up to 0.80.
	
	
2000-08-21  Sean M. Burke  sburke@cpan.org

	* Release 0.79

	* Bungled the Makefile on the previous release!  This time should work.
	
2000-08-21  Sean M. Burke  sburke@cpan.org

	* Release 0.78

	* Fixed odd bug in MIDI::Score that would prematurely terminate a
	pending note on one channel if you started (or ended) a note with
	the same note number on a different channel.

	New logic assumes (safely, I think) that you can only have one
	note at a time per note-number per channel.  (I.e., you can't
	start a C5 on channel 3, wait a second, start another C5 on
	channel 3, and have them both going at the same time.  I think
	that if you start a C5 on channel 3, then start another, the
	second will implicitly end the first.  Please do email me if you
	think this is non-standard behavior for a sequencer.)
	
2000-05-20  Sean M. Burke  sburke@cpan.org

	* Release 0.77

	* Fixing an incidental bug in MIDI::Simple (which surfaces only
	in Perl 5.6, apparently).  Thanks to Matt Burt (m.burt@bcs.org.uk)
	for finding the bug, as well as noting the fix.

	* Changed MIDI::Score::score_r_to_events_r and
	MIDI::Score::events_r_to_score_r (which were useful only in list
	context previously) to do something useful (and rather intuitive)
	in scalar context.
	
	
2000-05-14  Sean M. Burke  sburke@cpan.org

	* Release 0.76

        * Noting my new email address.
	
	* MIDI::_dump_quote() was a bit over-broad in what string values
	it considered numbers (i.e., what things it didn't have to put
	quotes around).  I've changed it to leave only integers unquoted.
	This shouldn't make any difference really, since if ever any Perl
	code (whether in MIDI-Perl or elsewhere) needs to treat a
	stringified numeral (whether integer or not) as a numeric value,
	conversion is automatic.  Anyhow, almost all (absolutely all?) of
	the actual numeric arguments to MIDI::Event events are integers.


2000-03-05  Sean M. Burke  sburke@netadventure.net

	* Release 0.75

	* ChangeLog now goes recent-to-old, not old-to-recent

	* Changes in this version were mostly made in August 1999, but
	just never released:

	* A few optimizations to Event.pm's loops.

	* Opus.pm now does a little more sanity-checking, to avoid
	attempting reads of insane sizes, such as might be attempted in
	the case of parsing a mangled file.

1999-05-13  Sean M. Burke  sburke@netadventure.net

	* Release 0.74

	* Just made a few things friendlier to people using perl -w
	(warnings)

	* Did ya see my article on MIDI-Perl in /The Perl Journal/ #14?
	See www.tpj.com for back issues, or email me -- I might
	have it in some sort of electronic format.

1999-01-10  Sean M. Burke  sburke@netadventure.net

	* Release 0.73

	* Just some changes to the docs:
	Changed the MIDI homepage URL.
	Added the Langston reference.

1998-11-09  Sean M. Burke  sburke@netadventure.net

	* Release 0.72

	* Shigeaki Kobayashi <xxkoby@pep.ne.jp> pointed out a bug in
	MIDI::Opus's $opus->draw that was making it always ignore its
	options hashref.  Fixed.
	
	* More carps/croaks instead of warn/dies.

1998-11-04  Sean M. Burke  sburke@netadventure.net

	* Release 0.71

	* Docs for MIDI::Simple are more complete now.
	
	* An obscure bit of note_map's behavior changed.
	But you'd never know, because this is the first version
	that ever documented note_map at all.

	* Added relative octave specs: o_d3, o_u3.

1998-10-18  Sean M. Burke  sburke@netadventure.net

	* Release 0.70
	
	* Some stupid typos in MIDI.pm fixed.

	* MIDI::Simple greatly expanded and changed.  Almost
	a total rewrite, in fact -- too many changes to explain.
	Docs are feeble, tho.

	* More modules use strict now.

1998-08-16  Sean M. Burke  sburke@netadventure.net

        * Release 0.62: Hooboy! Third release in a 24-hour period!
	This time to fix a lame big in MIDI::Score.

1998-08-16  Sean M. Burke  sburke@netadventure.net

	* Release 0.61: Feh.  Forgot to re-comment out the debug code
	in MIDI::Simple

1998-08-16  Sean M. Burke  sburke@netadventure.net

	* Release 0.60: 
	Subtly destructive bug in MIDI::_dump_quote fixed.
	
	MIDI::Event::copy_structure didn't work; fixed.
	
	%MIDI::note2number and %MIDI::note2number were named backwards --
	note2number actually mapped numbers to notes, and number2note
	mapped notes to numbers.  It made no sense, so I reversed it.

	Ditto patch2number and number2patch!  What was I thinking?

	Typos in %MIDI::note2number fixed.  All the A#'s were "A#" instead
	of "A#3" or whatever.  I also changed the "#" for sharp to "s",
	giving, e.g., "As3" instead of "A#3".  This gives a representation
	compatable with MIDI::Simple's notation.  But note that
	%MIDI::note2number doesn't contain all of the possible ways
	MIDI::Simple could let you represent a note -- just a
	(semi-quirky) subset of them.
	
	New module MIDI::Simple.
	

lib/MIDI/Score.pm

# Time-stamp: "2010-02-21 12:55:45 conklin"
require 5;
package MIDI::Score;
use strict;
use vars qw($Debug $VERSION);
use Carp;

$VERSION = '0.82';

=head1 NAME

MIDI::Score - MIDI scores

=head1 SYNOPSIS

  # it's a long story; see below

=head1 DESCRIPTION

This module provides functions to do with MIDI scores.
It is used as the basis for all the functions in MIDI::Simple.
(Incidentally, MIDI::Opus's draw() method also uses some of the
functions in here.)

Whereas the events in a MIDI event structure are items whose timing
is expressed in delta-times, the timing of items in a score is
expressed as an absolute number of ticks from the track's start time.
Moreover, pairs of 'note_on' and 'note_off' events in an event structure
are abstracted into a single 'note' item in a score structure.

'note' takes the following form:

 ('note_on', I<start_time>, I<duration>, I<channel>, I<note>, I<velocity>)

The problem that score structures are meant to solve is that 1)
people definitely don't think in delta-times -- they think in absolute
times or in structures based on that (like 'time from start of measure');
2) people think in notes, not note_on and note_off events.

So, given this event structure:

 ['text_event', 0, 'www.ely.anglican.org/parishes/camgsm/chimes.html'],
 ['text_event', 0, 'Lord through this hour/ be Thou our guide'],
 ['text_event', 0, 'so, by Thy power/ no foot shall slide'],
 ['patch_change', 0, 1, 8],
 ['note_on', 0, 1, 25, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 29, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 27, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 20, 96],
 ['note_off', 192, 0, 1, 0],
 ['note_on', 0, 1, 25, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 27, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 29, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 25, 96],
 ['note_off', 192, 0, 1, 0],
 ['note_on', 0, 1, 29, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 25, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 27, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 20, 96],
 ['note_off', 192, 0, 1, 0],
 ['note_on', 0, 1, 20, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 27, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 29, 96],
 ['note_off', 96, 0, 1, 0],
 ['note_on', 0, 1, 25, 96],
 ['note_off', 192, 0, 1, 0],

here is the corresponding score structure:

 ['text_event', 0, 'www.ely.anglican.org/parishes/camgsm/chimes.html'],
 ['text_event', 0, 'Lord through this hour/ be Thou our guide'],
 ['text_event', 0, 'so, by Thy power/ no foot shall slide'],
 ['patch_change', 0, 1, 8],
 ['note', 0, 96, 1, 25, 96],
 ['note', 96, 96, 1, 29, 96],
 ['note', 192, 96, 1, 27, 96],
 ['note', 288, 192, 1, 20, 96],
 ['note', 480, 96, 1, 25, 96],
 ['note', 576, 96, 1, 27, 96],
 ['note', 672, 96, 1, 29, 96],
 ['note', 768, 192, 1, 25, 96],
 ['note', 960, 96, 1, 29, 96],
 ['note', 1056, 96, 1, 25, 96],
 ['note', 1152, 96, 1, 27, 96],
 ['note', 1248, 192, 1, 20, 96],
 ['note', 1440, 96, 1, 20, 96],
 ['note', 1536, 96, 1, 27, 96],
 ['note', 1632, 96, 1, 29, 96],
 ['note', 1728, 192, 1, 25, 96]

Note also that scores aren't crucially ordered.  So this:

 ['note', 768, 192, 1, 25, 96],
 ['note', 960, 96, 1, 29, 96],
 ['note', 1056, 96, 1, 25, 96],

means the same thing as:

 ['note', 960, 96, 1, 29, 96],
 ['note', 768, 192, 1, 25, 96],
 ['note', 1056, 96, 1, 25, 96],

The only exception to this is in the case of things like:

 ['patch_change', 200,     2, 15],
 ['note',         200, 96, 2, 25, 96],

where two (or more) score items happen I<at the same time> and where one
affects the meaning of the other.

=head1 WHAT CAN BE IN A SCORE

Besides the new score structure item C<note> (covered above),
the possible contents of a score structure can be summarized thus:
Whatever can appear in an event structure can appear in a score
structure, save that its second parameter denotes not a
delta-time in ticks, but instead denotes the absolute number of ticks
from the start of the track.

To avoid the long periphrase "items in a score structure", I will
occasionally refer to items in a score structure as "notes", whether or
not they are actually C<note> commands.  This leaves "event" to
unambiguously denote items in an event structure.

These, below, are all the items that can appear in a score.
This is basically just a repetition of the table in
L<MIDI::Event>, with starttime substituting for dtime --
so refer to L<MIDI::Event> for an explanation of what the data types
(like "velocity" or "pitch_wheel").
As far as order, the first items are generally the most important:

=over

=item ('note', I<starttime>, I<duration>, I<channel>, I<note>, I<velocity>)

=item ('key_after_touch', I<starttime>, I<channel>, I<note>, I<velocity>)

=item ('control_change', I<starttime>, I<channel>, I<controller(0-127)>, I<value(0-127)>)

=item ('patch_change', I<starttime>, I<channel>, I<patch>)

=item ('channel_after_touch', I<starttime>, I<channel>, I<velocity>)

=item ('pitch_wheel_change', I<starttime>, I<channel>, I<pitch_wheel>)

=item ('set_sequence_number', I<starttime>, I<sequence>)

=item ('text_event', I<starttime>, I<text>)

=item ('copyright_text_event', I<starttime>, I<text>)

=item ('track_name', I<starttime>, I<text>)

=item ('instrument_name', I<starttime>, I<text>)

=item ('lyric', I<starttime>, I<text>)

=item ('marker', I<starttime>, I<text>)

=item ('cue_point', I<starttime>, I<text>)

=item ('text_event_08', I<starttime>, I<text>)

=item ('text_event_09', I<starttime>, I<text>)

=item ('text_event_0a', I<starttime>, I<text>)

=item ('text_event_0b', I<starttime>, I<text>)

=item ('text_event_0c', I<starttime>, I<text>)

=item ('text_event_0d', I<starttime>, I<text>)

=item ('text_event_0e', I<starttime>, I<text>)

=item ('text_event_0f', I<starttime>, I<text>)

=item ('end_track', I<starttime>)

=item ('set_tempo', I<starttime>, I<tempo>)

=item ('smpte_offset', I<starttime>, I<hr>, I<mn>, I<se>, I<fr>, I<ff>)

=item ('time_signature', I<starttime>, I<nn>, I<dd>, I<cc>, I<bb>)

=item ('key_signature', I<starttime>, I<sf>, I<mi>)

=item ('sequencer_specific', I<starttime>, I<raw>)

=item ('raw_meta_event', I<starttime>, I<command>(0-255), I<raw>)

=item ('sysex_f0', I<starttime>, I<raw>)

=item ('sysex_f7', I<starttime>, I<raw>)

=item ('song_position', I<starttime>)

=item ('song_select', I<starttime>, I<song_number>)

=item ('tune_request', I<starttime>)

=item ('raw_data', I<starttime>, I<raw>)

=back


=head1 FUNCTIONS

This module provides these functions:

=over

=item $score2_r = MIDI::Score::copy_structure($score_r)

This takes a I<reference> to a score structure, and returns a
I<reference> to a copy of it. Example usage:

          @new_score = @{ MIDI::Score::copy_structure( \@old_score ) };

=cut

sub copy_structure {
  return &MIDI::Event::copy_structure(@_);
  # hey, a LoL is an LoL
}
##########################################################################

=item $events_r = MIDI::Score::score_r_to_events_r( $score_r )

=item ($events_r, $ticks) = MIDI::Score::score_r_to_events_r( $score_r )

This takes a I<reference> to a score structure, and converts it to an
event structure, which it returns a I<reference> to.  In list context,
also returns a second value, a count of the number of ticks that
structure takes to play (i.e., the end-time of the temporally last
item).

=cut

sub score_r_to_events_r {
  # list context: Returns the events_r AND the total tick time
  # scalar context: Returns events_r
  my $score_r = $_[0];
  my $time = 0;
  my @events = ();
  croak "MIDI::Score::score_r_to_events_r's first arg must be a listref"
    unless ref($score_r);

  # First, turn instances of 'note' into 'note_on' and 'note_off':
  foreach my $note_r (@$score_r) {
    next unless ref $note_r;
    if($note_r->[0] eq 'note') {
      my @note_on = @$note_r;
#print "In:  ", map("<$_>", @note_on), "\n";
      $note_on[0] = 'note_on';
      my $duration = splice(@note_on, 2, 1);

      my @note_off = @note_on; # /now/ copy it
      $note_off[0] = 'note_off';
      $note_off[1] += $duration;
      $note_off[4] = 0; # set volume to 0
      push(@events, \@note_on, \@note_off);
#print "on:  ", map("<$_>", @note_on), "\n";
#print "off: ", map("<$_>", @note_off), "\n";
    } else {
      push(@events, [@$note_r]);
    }
  }
  # warn scalar(@events), " events in $score_r";
  $score_r = sort_score_r(\@events);
  # warn scalar(@$score_r), " events in $score_r";

  # Now we turn it into an event structure by fiddling the timing
  $time = 0;
  foreach my $event (@$score_r) {
    next unless ref($event) && @$event;
    my $delta =  $event->[1] - $time; # Figure out the delta
    $time = $event->[1]; # Move it forward
    $event->[1] = $delta; # Swap it in
  }
  return($score_r, $time) if wantarray;
  return $score_r;
}
###########################################################################

=item $score2_r = MIDI::Score::sort_score_r( $score_r)

This takes a I<reference> to a score structure, and returns a
I<reference> to a sorted (by time) copy of it. Example usage:

          @sorted_score = @{ MIDI::Score::sort_score_r( \@old_score ) };

=cut

sub sort_score_r {
  # take a reference to a score LoL, and sort it by note start time,
  # and return a reference to that sorted LoL.  Notes from the same
  # time must be left in the order they're found!!!!  That's why we can't
  # just use sort { $a->[1] <=> $b->[1] } (@$score_r)
  my $score_r = $_[0];
  my %timing = ();
  foreach my $note_r (@$score_r) {
    push(
	 @{$timing{
		   $note_r->[1]
		  }},
	 $note_r
	) if ref($note_r);
  }
# warn scalar(@$score_r), " events in $score_r";
#print "sequencing for times: ", map("<$_> ",
#				    sort {$a <=> $b} keys(%timing)
#				   ), "\n";

  return
    [
     map(@{ $timing{$_} },
	 sort {$a <=> $b} keys(%timing)
	)
    ];
}
###########################################################################

=item $score_r = MIDI::Score::events_r_to_score_r( $events_r )

=item ($score_r, $ticks) = MIDI::Score::events_r_to_score_r( $events_r )

This takes a I<reference> to an event structure, converts it to a
score structure, which it returns a I<reference> to.  If called in
list context, also returns a count of the number of ticks that
structure takes to play (i.e., the end-time of the temporally last
item).

=cut

sub events_r_to_score_r {
  # Returns the score_r AND the total tick time
  my $events_r = $_[0];
  croak "first argument to MIDI::Score::events_to_score is not a listref!"
    unless $events_r;
  my $options_r = ref($_[1]) ? $_[1] : {};

  my $time = 0;
  if( $options_r->{'no_note_abstraction'} ) {
    my $score_r = MIDI::Event::copy_structure($events_r);
    foreach my $event_r (@$score_r) {
      # print join(' ', @$event_r), "\n";
      $event_r->[1] = ($time += $event_r->[1]) if ref($event_r);
    }
    return($score_r, $time) if wantarray;
    return $score_r;
  } else {
    my %note = ();
    my @score =
      map
      {
	if(!ref($_)) {
	  ();
	} else {
# 0.82: the following must be declared local
	  local $_ = [@$_]; # copy.
	  $_->[1] = ($time += $_->[1]) if ref($_);
	  
	  if($_->[0] eq 'note_off'
	     or($_->[0] eq 'note_on' &&
		$_->[4] == 0) )
	  { # End of a note
	    # print "Note off : @$_\n";
# 0.82: handle multiple prior events with same chan/note.
	      if ((exists $note{pack 'CC', @{$_}[2,3]}) && (@{$note{pack 'CC', @{$_}[2,3]}})) {
		  shift(@{$note{pack 'CC', @{$_}[2,3]}})->[2] += $time;
		  unless(@{$note{pack 'CC', @{$_}[2,3]}}) {delete $note{pack 'CC', @{$_}[2,3]};}
	      }
	    (); # Erase this event.
	  } elsif ($_->[0] eq 'note_on') {
	    # Start of a note
	    $_ = [@$_];
	    
	    push(@{$note{ pack 'CC', @{$_}[2,3] }},$_);
	    splice(@$_, 2, 0, -$time);
	    $_->[0] = 'note';
	    # ('note', Starttime, Duration, Channel, Note, Veloc)
	    $_;
	  } else {
	    $_;
	  }
	}
      }
      @$events_r
    ;

    #print "notes remaining on stack: ", scalar(values %note), "\n"
    #  if values %note;
# 0.82: clean up pending events gracefully
    foreach my $k (keys %note) {
	foreach my $one (@{$note{$k}}) {
	    $one->[2] += $time;
	}
    }
    return(\@score, $time) if wantarray;
    return \@score;
  }
}
###########################################################################

=item $ticks = MIDI::Score::score_r_time( $score_r )

This takes a I<reference> to a score structure, and returns 
a count of the number of ticks that structure takes to play
(i.e., the end-time of the temporally last item).

=cut

sub score_r_time {
  # returns the duration of the score you pass a reference to
  my $score_r = $_[0];
  croak "arg 1 of MIDI::Score::score_r_time isn't a ref" unless ref $score_r;
  my $track_time = 0;
  foreach my $event_r (@$score_r) {
    next unless @$event_r;
    my $event_end_time = ($event_r->[0] eq 'note') ?
      ($event_r->[1] + $event_r->[2])  :  $event_r->[1] ;
    #print "event_end_time: $event_end_time\n";
    $track_time = $event_end_time if $event_end_time > $track_time;
  }
  return $track_time;
}
###########################################################################

=item MIDI::Score::dump_score( $score_r )

This dumps (via C<print>) a text representation of the contents of
the event structure you pass a reference to.

=cut

sub dump_score {
  my $score_r = $_[0];
  print "\@notes = (   # ", scalar(@$score_r), " notes...\n";
  foreach my $note_r (@$score_r) {
    print " [", &MIDI::_dump_quote(@$note_r), "],\n" if @$note_r;
  }
  print ");\n";
  return;
}

###########################################################################

=item MIDI::Score::quantize( $score_r )

This takes a I<reference> to a score structure, performs a grid
quantize on all events, returning a new score reference with new
quantized events.  Two parameters to the method are: 'grid': the
quantization grid, and 'durations': whether or not to also quantize
event durations (default off).

When durations of note events are quantized, they can get 0 duration.
These events are I<not dropped> from the returned score, and it is the
responsiblity of the caller to deal with them.

=cut

# new in 0.82!
sub quantize {
  my $score_r = $_[0];
  my $options_r = ref($_[1]) eq 'HASH' ? $_[1] : {};
  my $grid = $options_r->{grid};
  if ($grid < 1) {carp "bad grid $grid in MIDI::Score::quantize!"; $grid = 1;}
  my $qd = $options_r->{durations}; # quantize durations?
  my $new_score_r = [];
  my $n_event_r;
  foreach my $event_r (@{$score_r}) {
      my $n_event_r = [];
      @{$n_event_r} = @{$event_r};
      $n_event_r->[1] = $grid * int(($n_event_r->[1] / $grid) + 0.5);
      if ($qd && $n_event_r->[0] eq 'note') {
	  $n_event_r->[2] = $grid * int(($n_event_r->[2] / $grid) + 0.5);
      }
      push @{$new_score_r}, $n_event_r;
  }
  $new_score_r;
}


###########################################################################

=back

=head1 COPYRIGHT 

Copyright (c) 1998-2002 Sean M. Burke. All rights reserved.

This library is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 AUTHORS

Sean M. Burke C<sburke@cpan.org> (until 2010)

Darrell Conklin C<conklin@cpan.org> (from 2010)

=cut

1;

__END__

